In [1]:
import xarray as xr
import pandas as pd
import numpy as np
from dask.distributed import Client

In [2]:
# qsub -IVX -l nodes=1:ppn=10,mem=40g,vmem=40g,walltime=01:00:00
# client = Client(n_workers=10, threads_per_worker=1, memory_limit='4GB')
client = Client()
client

<Client: 'tcp://127.0.0.1:45141' processes=5 threads=10, memory=42.95 GB>

In [10]:
client.close()

In [3]:
da = xr.open_mfdataset('/bdd/ERA5/NETCDF/GLOBAL_025/4xdaily/AN_PL/2020/w.2020*.aphe5.GLOBAL_025.nc', parallel=True, chunks={'latitude':100, 'longitude':100}).w
da

<xarray.DataArray 'w' (time: 1464, level: 37, latitude: 721, longitude: 1440)>
dask.array<concatenate, shape=(1464, 37, 721, 1440), dtype=float32, chunksize=(124, 37, 100, 100), chunktype=numpy.ndarray>
Coordinates:
  * latitude   (latitude) float32 90.0 89.75 89.5 89.25 ... -89.5 -89.75 -90.0
  * longitude  (longitude) float32 0.0 0.25 0.5 0.75 ... 359.25 359.5 359.75
  * level      (level) int32 1 2 3 5 7 10 20 30 ... 850 875 900 925 950 975 1000
  * time       (time) datetime64[ns] 2020-01-01 ... 2020-12-31T18:00:00
Attributes:
    units:          Pa s**-1
    long_name:      Vertical velocity
    standard_name:  lagrangian_tendency_of_air_pressure
    level_desc:     pressure level

- With automatic chunks (on time)
Array	Chunk  
Bytes	224.96 GB	19.05 GB  
Shape	(1464, 37, 721, 1440)	(124, 37, 721, 1440)  
Count	36 Tasks	12 Chunks  
Type	float32	numpy.ndarray  

- With additional chunks over lat/lon
Array	Chunk  
Bytes	224.96 GB	183.52 MB  
Shape	(1464, 37, 721, 1440)	(124, 37, 100, 100)  
Count	2892 Tasks	1440 Chunks  
Type	float32	numpy.ndarray  

In [4]:
# Generate random lat/lon
lats = np.random.uniform(low=-90, high=90, size=(200,))
lons = np.random.uniform(low=0, high=360, size=(200,))

In [5]:
da_latlon = da.sel(latitude=lats, longitude=lons, method='nearest')
da_latlon

/data/mlalande/miniconda3/envs/phd_v3/lib/python3.8/site-packages/xarray/core/indexing.py:1349: PerformanceWarning: Slicing with an out-of-order index is generating 13 times more chunks
  value = value[(slice(None),) * axis + (subkey,)]
/data/mlalande/miniconda3/envs/phd_v3/lib/python3.8/site-packages/xarray/core/indexing.py:1349: PerformanceWarning: Slicing with an out-of-order index is generating 22 times more chunks
  value = value[(slice(None),) * axis + (subkey,)]


<xarray.DataArray 'w' (time: 1464, level: 37, latitude: 200, longitude: 200)>
dask.array<getitem, shape=(1464, 37, 200, 200), dtype=float32, chunksize=(124, 37, 3, 4), chunktype=numpy.ndarray>
Coordinates:
  * latitude   (latitude) float32 -16.0 22.75 17.5 53.25 ... -44.75 49.75 -13.75
  * longitude  (longitude) float32 337.5 52.75 199.0 67.5 ... 263.75 71.5 100.0
  * level      (level) int32 1 2 3 5 7 10 20 30 ... 850 875 900 925 950 975 1000
  * time       (time) datetime64[ns] 2020-01-01 ... 2020-12-31T18:00:00
Attributes:
    units:          Pa s**-1
    long_name:      Vertical velocity
    standard_name:  lagrangian_tendency_of_air_pressure
    level_desc:     pressure level

Array	Chunk  
Bytes	8.67 GB	734.08 MB  
Shape	(1464, 37, 200, 200)	(124, 37, 200, 200)  
Count	60 Tasks	12 Chunks  
Type	float32	numpy.ndarray  

In [8]:
%%time
da_latlon.load()


KeyboardInterrupt

distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker process still alive after 3 seconds, killing
distributed.nanny - WARNING - Worker process still alive after 3 seconds, killing


In [ ]:
da_latlon.chunk({'time':-1, 'latitude':50, 'longitude':50, 'levels':10})

In [ ]:
# Interpolate to 3h
da_latlon_interp = da_latlon.interp(time=pd.date_range(da.time[0].values, da.time[-1].values, freq="3H"))

In [6]:
da.time[0]

<xarray.DataArray 'time' ()>
array('2020-01-01T00:00:00.000000000', dtype='datetime64[ns]')
Coordinates:
    time     datetime64[ns] 2020-01-01
Attributes:
    long_name:  time

In [8]:
pd.date_range(da.time[0].values, da.time[-1].values, freq="3H")

DatetimeIndex(['2020-01-01 00:00:00', '2020-01-01 03:00:00',
               '2020-01-01 06:00:00', '2020-01-01 09:00:00',
               '2020-01-01 12:00:00', '2020-01-01 15:00:00',
               '2020-01-01 18:00:00', '2020-01-01 21:00:00',
               '2020-01-02 00:00:00', '2020-01-02 03:00:00',
               ...
               '2020-12-30 15:00:00', '2020-12-30 18:00:00',
               '2020-12-30 21:00:00', '2020-12-31 00:00:00',
               '2020-12-31 03:00:00', '2020-12-31 06:00:00',
               '2020-12-31 09:00:00', '2020-12-31 12:00:00',
               '2020-12-31 15:00:00', '2020-12-31 18:00:00'],
              dtype='datetime64[ns]', length=2927, freq='3H')

In [ ]:
da_latlon.interp(time=)

In [6]:
%%time
da.interp(longitude=20.1, latitude=10.1, time='2020-02-12T18:00:00', method='linear').values

distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting
distributed.nanny - WARNING - Restarting worker
distributed.nann

KeyboardInterrupt: 

distributed.nanny - WARNING - Restarting worker


In [ ]:
%%time
da.interp(longitude=20.1, latitude=10.1, time='2020-02-12T18:00:00', method='linear').values